driver/main.c

```diff
@@ -720,98 +591,172 @@ static int ppm_release(struct inode *inode, struct file *filp)
 	 * The last closed device stops event collection
 	 */
 	if (list_empty(&g_consumer_list)) {
-		if (g_tracepoint_registered) {
+		if (g_tracepoints_attached != 0) {
 			pr_info("no more consumers, stopping capture\n");
 
-			/* 
-			 * SYS_EXIT 
+			force_tp_set(0, TP_VAL_MAX);
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 20)
+			tracepoint_synchronize_unregister();
+#endif
+			g_tracepoints_attached = 0;
+
+			/*
+			 * Reset tracepoint counter
 			 */
+			for_each_possible_cpu(cpu) {
+				per_cpu(g_n_tracepoint_hit, cpu) = 0;
+			}
+		} else {
+			ASSERT(false);
+		}
+	}
+
+	ret = 0;
+
+cleanup_release:
+	mutex_unlock(&g_consumer_mutex);
+
+	return ret;
+}
+
+static int compat_set_tracepoint(void *func, const char *probename, struct tracepoint *tp, bool enabled)
+{
+	int ret = 0;
+	if (enabled)
+	{
+		ret = compat_register_trace(func, probename, tp);
+	}
+	else
+	{
+		compat_unregister_trace(func, probename, tp);
+	}
+	return ret;
+}
+
+static int force_tp_set(u32 new_tp_set, u32 max_val)
+{
+	u32 idx;
+	u32 new_val;
+	u32 curr_val;
+	int ret = 0;
+
+	if (new_tp_set == g_tracepoints_attached)
+	{
+		// ok already equal
+		return ret;
+	}
+
+	for(idx = 0; idx < max_val && ret == 0; idx++)
+	{
+		ret = 0;
```

Comment:
We probably don't need this, since `ret` is initialized to 0 and the for loop stops when its value changes.